### PR TITLE
Write an architectural overview of ci-operator

### PR DESCRIPTION
**Pull request by stevekuznetsov**
Signed-off-by: Steve Kuznetsov <skuznets@redhat.com>

/assign @petr-muller @droslean @bbguimaraes

### CODE REVIEW

File: ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+
+In order to achieve this, the CI Operator requires configuration to understand
+the build process for each component as well as every output container image.
+This document will overview the workflow that the CI Operator uses to build
+components and structure tests.
+
+Every invocation of `ci-operator` will create a workspace to isolate test
+execution, seed it with build inputs and the published component images from
+the OpenShift release that the test is a part of, then schedule test workflows
+as Kubernetes and OpenShift objects. This workflow is executed using the
```
Comment (bbguimaraes) [inline ARCHITECTURE.md]: `tag_specification` is optional, we have a few repositories that do not use it.

File: ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+All input images for the tests that are described in the configuration YAML are
+tagged in, as are all images that form the larger release that the test is a part
+of. Images that are used for the build graph, like those identified with the
+[`base_images`](./CONFIGURATION.md#base_images),
+[`base_rpm_images`](./CONFIGURATION.md#base_rpm_images), and
+[`build_root`](./CONFIGURATION.md#build_root) stanzas, have `ImageStreamTag`s
+created for them in the `pipeline` `ImageStream` in the test `Namespace`. Images
+that are part of the release that the test exists within, as specified with the
+[`tag_specification`](./CONFIGURATION.md#tag_specification) stanza, are mirrored
+to `ImageStreamTag`s in the `stable` `ImageStream` within the test `Namespace`.
```
Comment (bbguimaraes) [inline ARCHITECTURE.md]: Same here (twice).

File: ARCHITECTURE.md [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+
+In order to ensure that resources from tests do not leak to the cluster the tests
+are executed on, both hard and soft TTLs are set on the `Namespace` and the
+[`ci-ns-ttl-controller`](https://github.com/openshift/ci-ns-ttl-controller) is
+used to enforce the TTLs and reap namespaces when TTLs have expired.
```
Comment (bbguimaraes) [inline ARCHITECTURE.md]: We've been asked a few times about this, so it's worth expanding this section to say that the TTL controller is deployed on `api.ci` and describe what happens (or doesn't) when it's not deployed or if the user doesn't have permission to edit the namespace.

File: ARCHITECTURE.md [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+Container image builds -- whether from the implicit `pipeline` steps above or
+from explicit image build configurations in the [`images`](./CONFIGURATION.md#images)
+stanza, are executed using OpenShift `Build`s. Test targets in the [`tests`](./CONFIGURATION.md#tests)
+stanza are executed using Kubernetes `Pod`s. As all of the test workflow execution
+objects are created in a `Namespace` shared for all jobs with the same input,
+re-use is achieved by deterministic naming. For instance, the `src` `Build` that
+creates the `pipeline:src` `ImageStreamTag` will be created only once in a given
+`Namespace`; other builds of jobs that require this build step will see the
+`Build` running and wait for it to complete or see the `ImageStreamTag` existing
+and consider the build step finished.
```
Comment (bbguimaraes) [inline ARCHITECTURE.md]: One thing missing from all our documentation files that I was hoping to see here is the different types of image builds that are supported and the corresponding `ci-operator` configuration entry. A few examples extracted quickly from the existing configuration files:

```yaml
- dockerfile_path: test/install/Dockerfile.Fedora
  from: bin
  to: fedora-install
- dockerfile_path: images/installer/Dockerfile.ci
  from: base
  inputs:
    root:
      as:
      - build
  to: installer
- from: os
  inputs:
    root:
      as:
      - '0'
  to: release-controller
- dockerfile_path: Dockerfile.e2e
  from: base
  inputs:
    bin:
      paths:
      - destination_dir: .
        source_path: /go/src/github.com/openshift/openshift-azure/e2e.test
    root:
      as:
      - e2e-tests
  to: e2e-tests
```

File: ARCHITECTURE.md [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+ - hide complexity from the end user
+
+In order to achieve this, the CI Operator requires configuration to understand
+the build process for each component as well as every output container image.
+This document will overview the workflow that the CI Operator uses to build
+components and structure tests.
+
+Every invocation of `ci-operator` will create a workspace to isolate test
+execution, seed it with build inputs and the published component images from
+the OpenShift release that the test is a part of, then schedule test workflows
```
Comment (petr-muller) [inline ARCHITECTURE.md]: '...the test is a part of...' - does this make sense? the test is probably not part of the release, a tested component is?

File: ARCHITECTURE.md [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+of. Images that are used for the build graph, like those identified with the
+[`base_images`](./CONFIGURATION.md#base_images),
+[`base_rpm_images`](./CONFIGURATION.md#base_rpm_images), and
+[`build_root`](./CONFIGURATION.md#build_root) stanzas, have `ImageStreamTag`s
+created for them in the `pipeline` `ImageStream` in the test `Namespace`. Images
+that are part of the release that the test exists within, as specified with the
+[`tag_specification`](./CONFIGURATION.md#tag_specification) stanza, are mirrored
+to `ImageStreamTag`s in the `stable` `ImageStream` within the test `Namespace`.
+
+In order to ensure that resources from tests do not leak to the cluster the tests
```
Comment (petr-muller) [inline ARCHITECTURE.md]: in order to

also, "leak **to** the cluster"? What does that mean?

File: ARCHITECTURE.md (hunk 4 above)
Comment (petr-muller) [reply ARCHITECTURE.md]: Me too, but I wonder if architecture description is the right place to put these. I start to think we might need a cookbook with examples - not just for images but also for tests, especially now with the new test types.

Comment (stevekuznetsov) [reply ARCHITECTURE.md]: The configuration doc points to the OpenShift Build documents -- this is just using OpenShift Build flows

**Comment by stevekuznetsov** [comment]
@bbguimaraes @petr-muller updated

File: ARCHITECTURE.md (hunk 4 above)
Comment (bbguimaraes) [reply ARCHITECTURE.md]: /shrug

**Pull request merged**